python/lsst/ts/ess/csc/controller_data_client.py

```diff
@@ -0,0 +1,449 @@
+            raise RuntimeError(
+                f"Could not parse {read_bytes!r} as a json-encoded dict."
+            )
+        return data
+
+    async def run_command(self, command: str, **parameters: Any) -> None:
+        """Write a command. Time out if it takes too long.
+
+        Parameters
+        ----------
+        command : `str`
+            The command to write.
+        **parameters : `dict`
+            Command parameters, as name=dict. For example::
+
+                configuration = {"devices": self.config.devices}
+
+        Raises
+        ------
+        ConnectionError
+            If not connected
+        asyncio.TimeoutError
+            If it takes more than COMMUNICATE_TIMEOUT seconds
+            to acquire the lock or write the data.
+        """
+        json_str = json.dumps({"command": command, "parameters": parameters})
+        await asyncio.wait_for(
+            self._basic_run_command(json_str), timeout=COMMUNICATE_TIMEOUT
+        )
+
+    async def _basic_run_command(self, json_str: str) -> None:
+        """Write a json-encoded command dict. Potentially wait forever.
+
+        Parameters
+        ----------
+        json_str : `str`
+            json-encoded dict to write. The dict should be of the form::
+
+                {"command": command_str, "parameters": params_dict}
+
+        Raises
+        ------
+        RuntimeError
+            If the command fails.
+        ConnectionError
+            If disconnected before command is acknowledged.
+        """
+        async with self.stream_lock:
+            if not self.connected:
+                raise ConnectionError("Not connected; cannot send the command.")
+            assert self.writer is not None  # make mypy happy
+
+            self.writer.write(json_str.encode() + tcpip.TERMINATOR)
+            await self.writer.drain()
+            while True:
+                if not self.connected:
+                    raise ConnectionError(
+                        "Disconnected while waiting for command response"
+                    )
+                data = await self.read()
+                if common.Key.RESPONSE in data:
+                    response = data[common.Key.RESPONSE]
+                    if response == common.ResponseCode.OK:
+                        return
+                    else:
+                        raise RuntimeError(f"Command {json_str!r} failed: {response!r}")
+                else:
+                    self.log.debug("Ignoring non-command-ack")
+
+    async def process_telemetry(
+        self,
+        sensor_name: str,
+        timestamp: float,
+        response_code: int,
+        sensor_data: Sequence[float | int | str],
+    ) -> None:
+        """Process the sensor telemetry
+
+        Parameters
+        ----------
+        sensor_name : `str`
+            The name of the sensor.
+        timestamp : `float`
+            The timestamp of the data.
+        response_code : `int`
+            The ResponseCode
```

Comment:
Is this doc string done?

Reply:
It is. I added a period to indicate that.